mesa/space.py

```diff
@@ -849,18 +849,15 @@ def __init__(
     def _build_agent_cache(self):
         """Cache agents positions to speed up neighbors calculations."""
         self._index_to_agent = {}
-        agents = self._agent_to_index.keys()
-        for idx, agent in enumerate(agents):
+        for idx, agent in enumerate(self._agent_to_index):
             self._agent_to_index[agent] = idx
             self._index_to_agent[idx] = agent
-        self._agent_points = np.array(
-            [self._index_to_agent[idx].pos for idx in range(len(agents))]
-        )
+        # Since dicts are ordered by insertion, we can iterate through agents keys
+        self._agent_points = np.array([agent.pos for agent in self._agent_to_index])
 
     def _invalidate_agent_cache(self):
         """Clear cached data of agents and positions in the space."""
         self._agent_points = None
-        self._index_to_agent = {}
```

Comment:
This attribute is used in `get_neighbors`, which hasn't been refactored away.

Reply:
The cache clearing functionality should not depend on how the cache is checked/rebuilt. `self._index_to_agent` is a cache that should be cleared in this method regardless of how it is used elsewhere.

Reply:
Understood 👍